Use a single ThreadPool for "spack versions"

alalazo opened the pull request:
fixes #16720 

Modifications:

- [x] `spack versions` doesn't spawn recursive pool of processes anymore
- [x]  The number of concurrent requests can be controlled via cli

alalazo committed:
Serialized web spidering

alalazo committed:
Using a single thread pool for spidering

The spider function uses a single ThreadPool
instead of nested ProcessPool objects.

The implementation of spidering for a
single url is now a nested function.
This permits to capture a local cache
of visited links.

Updated docstrings

alalazo committed:
Added a "concurrency" argument to the spider function

alalazo committed:
Control the number of threads spawned by "spack version" from cli

alalazo commented on lib/spack/spack/package.py:
```diff
@@ -2020,7 +2020,7 @@ def all_urls(self):
         return urls
 
-    def fetch_remote_versions(self):
+    def fetch_remote_versions(self, concurrency=128):
```
This magic number works well on my development box, but I wonder if we want to use different defaults for different contexts (e.g. `128` for cli argument and a much lower value for the API)

tgamblin replied:
I would pick the lowest number you can use without impacting the speed of fetching.  Like pick a package with a fairly deep tree to spider and see how much 128 really improves things vs 64 or 32.  There is probably a knee in that curve

alalazo replied:
That's what I did last time for:
```console
$ time spack versions -c X  boost
```
on my Lenovo Carbon laptop. Repeating the experiment today, the results changed a bit and 128 is not anymore the clear winner:

Threads | Elapsed time
------------|------------------
16 | 124 sec.
32 | 69 sec.
64 | 56 sec.
128 | 53 sec.

I expect the time it takes for this command to run would be highly inconsistent and to depend on the current latency / flakiness of the network. For instance re-doing the run with 64 threads a few times I got an "outlier" at 93 sec. That said, I can change the default to 32. Let me know what you prefer here.

alalazo replied:
Changed default to 32 in abd85f9

alalazo commented:
@nazavode

adamjstewart commented:
> 81.32% of diff hit (target 81.44%)

So close, yet so far...

alalazo committed:
test/web.py: rewrite a few from imports as regular imports

alalazo committed:
test/web.py: unified spider unit tests

alalazo committed:
Added a new unit test for no response on spidering

alalazo commented on lib/spack/spack/util/web.py:
```diff
@@ -336,161 +309,194 @@ def list_url(url):
 
-        if raise_on_error:
-            raise NoNetworkConnectionError(str(e), url)
+                # Skip things outside the root directory
+                if not abs_link.startswith(root):
+                    continue
```
@tgamblin @adamjstewart When debugging I noticed that here we are testing if `abs_link` (a string) starts with a `ParseResult` object, which will always be false. Do we want to fix the logic, or delete the test (and thus keep the current actual behavior)?

adamjstewart replied:
If it's always false, I would delete the check to speed things up. But I would need to take a deeper look at the code to make sure it isn't needed.

tgamblin replied:
The intent here was to avoid leaving the root subdirectories (e.g. going to entirely different websites), so fixing the check would likely speed things up by not exploring links that go off the root site. 

It might be worth running `spack versions` for each package and seeing how much this helps and whether it affects the output.  I can't decide what to do because there might be packages where spelunking off the root site is needed.  I kind of doubt it, but fixing this would make it so we couldn't get to those pages.  Would it affect github releases?  Do those links go outside the root page?  If so then we should probably delete the check.

alalazo replied:
I just checked: fixing the check makes 15 tests fail and several packages are finding less versions (e.g. `boost`). I'll delete the check for now, which means I'll keep the current behavior. Let me know if you want me to open an issue to revisit this logic later and optimize the search.

alalazo replied:
See e85b6fe and its comment for a broader explanation of what's going on.

luigi-calori commented:
Confirm that on our Power9 system this PR solves the issue #16720 that wac causing node 
hangs when issueing
spack versions openfoam

tgamblin commented on lib/spack/spack/cmd/versions.py:
```diff
@@ -21,6 +21,10 @@
                            help='only list safe versions of the package')
+    subparser.add_argument(
+        '-c', '--concurrency', default=128,
+        help='number of concurrent requests'
```
add a kwarg `type=int` here

tgamblin commented on lib/spack/spack/cmd/versions.py:
```diff
@@ -45,7 +49,7 @@ def versions(parser, args):
         tty.msg('Remote versions (not yet checksummed):')
 
-    fetched_versions = pkg.fetch_remote_versions()
+    fetched_versions = pkg.fetch_remote_versions(int(args.concurrency))
```
if you add a kwarg to `add_argument` above you don't need `int()` here

tgamblin committed:
swap order

tgamblin commented:
@alalazo: I moved urlopen to after spider to make the diff intelligible (since the function wasn't modified)

tgamblin commented on lib/spack/spack/util/web.py:
```diff
@@ -336,107 +309,155 @@ def list_url(url):
 
-def _spider(url, visited, root, depth, max_depth, raise_on_error):
-    """Fetches URL and any pages it links to up to max_depth.
+def spider(roots, depth=0, concurrency=128):
```
rename roots to `root_urls`

alalazo committed:
spack versions: do int conversion using argparse API

alalazo committed:
Renamed root to root_urls

alalazo committed:
Remove the check on root url

The current check is buggy and the conditional always
evaluate to False. What happens is that the current
url is checked against a ParseResult object. This, in
turn, is seen as a tuple of strings. Since some of the
strings in the tuple are empty "url.startswith(root)"
always evaluate to True and the check always evaluate
to False.

Here we remove the check to keep the current visible
behavior, since "fixing" it makes 15 unit tests fail.

alalazo committed:
Make 32 (instead of 128) the default number of concurrent requests
